### ISSUE TITLE
Support shared_preload_libraries etc...

### ISSUE THREAD

**Issue by jamessewell** [bluejekyll/pg-extend-rs#45]
Hi,

Just wanted to know how you'd feel about adding support for `shared_preload_libraries` / `bgworkers` \ shared memory (although that's a lot harder) being added?

**Comment by bluejekyll** [comment]
I’m a big fan of us using this library for extending Postgres in as many different ways as possible.

I’m currently a little concerned about the memory model for a lot of this, but yes, I’m a big supporter.

**Comment by jamessewell** [comment]
What are the concerns? I see you’re using palloc in behind a feature gate -
which is neat. I also saw a comment about state being lost - but that’s
what shared memory is for.

Admittedly writing to Shared Memory is a bit iffy in my experience - that
can be worked on though.

**Comment by bluejekyll** [comment]
I’m working on a patch to try and make things threadsafe in regards to the use of palloc. I can reproduce issues with it right now, but don’t have it isolated to why the exact problem is, yet.

**Comment by jamessewell** [comment]
I’m struggling to get my head round a usecase for threading without loading
as a shared library?

Are you on Slack or irc anywhere?

**Comment by bluejekyll** [comment]
I hang out in the Rust discord app, we can chat there.

This is all in the context of the shared library loading, as currently implemented. Threading would be from the Rust, though it might be in appropriate in general in Postgres, but I expect people coming from Rust will want to use threading...

**Comment by jamessewell** [comment]
Just saw this issue I'd forgotten and thought I'd comment with the current state of play - turns out calling Postgres through FFI (as this crate does) is unsafe even *in the presence of threads*.

ie: if you use Rayon, or something which starts a thread to do work in the background then you are firmly into the realms of UB - even if the threads never call into Postgres.

**Comment by bluejekyll** [comment]
I agree that threads with PG FFI are inherently dangerous, given the nature of PG.

I think an option we have is to pass around a `& mut` instance of the memory context, and that would not be sendable across threads. If we required that handle on all FFI, then it would be possible to reduce mistakes in that context.

Not sure it’s worth it though. BG workers are probably the safe and best model here.

**Comment by jamessewell** [comment]
The issue with that approach is usually just because something can’t be
sent between threads doesn’t mean it can’t be created on another thread.

Bgworkers suffer the same fate, with the small proviso that they can thread
if they never connect to SPI, call PG functions, or deal with signals.

**Comment by bluejekyll** [comment]
So we already mark `PgDatum` and `PgAllocated` types as both `!Send` and `!Sync` specifically to deny sharing memory managed by PG between threads.

Otherwise we use the default Rust allocator for all memory. Is there a reason you think that isn't safe? I'm not honestly sure we have any changes necessary with those restrictions in place.

**Comment by jamessewell** [comment]
The handle on FFI idea is interesting - even things like postgresql logging
from a thread are unsafe sadly.
